contracts/crosschain/ERC7786Aggregator.sol

```diff
@@ -0,0 +1,278 @@
+        for (uint256 i = 0; i < _gateways.length(); ++i)
+            if (!IERC7786GatewaySource(_gateways.at(i)).supportsAttribute(selector)) return false;
+        return true;
+    }
+
+    /// @inheritdoc IERC7786GatewaySource
+    function sendMessage(
+        string calldata destinationChain,
+        string memory receiver, // using memory instead of calldata avoids stack too deep error
+        bytes memory payload, // using memory instead of calldata avoids stack too deep error
+        bytes[] memory attributes // using memory instead of calldata avoids stack too deep error
+    ) external payable whenNotPaused returns (bytes32 outboxId) {
+        if (msg.value > 0) revert ERC7786AggregatorValueNotSupported();
+        // address of the remote router, revert if not registered
+        string memory router = getRemoteRouter(destinationChain);
+
+        // wrapping the payload
+        bytes memory wrappedPayload = abi.encode(++_nonce, msg.sender.toChecksumHexString(), receiver, payload);
+
+        // Post on all gateways
+        Outbox[] memory outbox = new Outbox[](_gateways.length());
+        bool needsId = false;
+        for (uint256 i = 0; i < outbox.length; ++i) {
+            address gateway = _gateways.at(i);
+            // send message
+            bytes32 id = IERC7786GatewaySource(gateway).sendMessage(
+                destinationChain,
+                router,
+                wrappedPayload,
+                attributes
+            );
+            // if ID, track it
+            if (id != bytes32(0)) {
+                outbox[i] = Outbox(gateway, id);
+                needsId = true;
+            }
+        }
+
+        if (needsId) {
+            outboxId = keccak256(abi.encode(outbox));
+            emit OutboxDetails(outboxId, outbox);
+        }
+
+        emit MessagePosted(
+            outboxId,
+            CAIP10.local(msg.sender),
+            CAIP10.format(destinationChain, receiver),
+            payload,
+            attributes
+        );
+    }
+
+    // ============================================== IERC7786Receiver ===============================================
+
+    /// @inheritdoc IERC7786Receiver
+    function executeMessage(
+        string calldata sourceChain, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address (does not include the chain identifier)
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable whenNotPaused returns (bytes4) {
+        // Check sender is a trusted remote router
+        if (!_remoteRouters[sourceChain].equal(sender)) revert ERC7786AggregatorInvalidCrosschainSender();
+
+        // Message reception tracker
+        bytes32 id = keccak256(abi.encode(sourceChain, sender, payload, attributes));
+        Tracker storage tracker = _trackers[id];
+
+        // If call originates from a trusted gateway
+        if (_gateways.contains(msg.sender) && !tracker.receivedBy[msg.sender]) {
```

Comment:
Any reason not to revert if `msg.sender` is not a trusted gateway?

Reply:
See https://github.com/OpenZeppelin/openzeppelin-community-contracts/pull/82#discussion_r1985292425